{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ffd31ea-b652-4721-84c8-a7f60ea5394b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-11-23 21:16:14--  https://alphafold.ebi.ac.uk/files/AF-P04637-F1-model_v4.cif\n",
      "Resolving alphafold.ebi.ac.uk (alphafold.ebi.ac.uk)... 34.149.152.8\n",
      "Connecting to alphafold.ebi.ac.uk (alphafold.ebi.ac.uk)|34.149.152.8|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: unspecified [application/octet-stream]\n",
      "Saving to: ‘AF-P04637-F1-model_v4.cif’\n",
      "\n",
      "AF-P04637-F1-model_     [ <=>                ] 351.36K  --.-KB/s    in 0.04s   \n",
      "\n",
      "2023-11-23 21:16:15 (7.92 MB/s) - ‘AF-P04637-F1-model_v4.cif’ saved [359793]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://alphafold.ebi.ac.uk/files/AF-P04637-F1-model_v4.cif"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8cec94b6-fe76-4b6e-bc93-88d652f852a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting biopython\n",
      "  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m18.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0ma \u001b[36m0:00:01\u001b[0m\n",
      "\u001b[?25hCollecting nglview\n",
      "  Downloading nglview-3.0.8.tar.gz (6.8 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m6.8/6.8 MB\u001b[0m \u001b[31m13.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: ipywidgets in /opt/mamba/lib/python3.10/site-packages (8.1.1)\n",
      "Requirement already satisfied: numpy in /opt/mamba/lib/python3.10/site-packages (from biopython) (1.26.0)\n",
      "Requirement already satisfied: jupyterlab-widgets in /opt/mamba/lib/python3.10/site-packages (from nglview) (3.0.9)\n",
      "Requirement already satisfied: comm>=0.1.3 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets) (0.1.4)\n",
      "Requirement already satisfied: ipython>=6.1.0 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets) (8.17.2)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets) (5.13.0)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.9 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets) (4.0.9)\n",
      "Requirement already satisfied: decorator in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (0.19.1)\n",
      "Requirement already satisfied: matplotlib-inline in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (0.1.6)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (3.0.41)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (2.17.0)\n",
      "Requirement already satisfied: stack-data in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (0.6.2)\n",
      "Requirement already satisfied: exceptiongroup in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (1.1.3)\n",
      "Requirement already satisfied: pexpect>4.3 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets) (4.8.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /opt/mamba/lib/python3.10/site-packages (from jedi>=0.16->ipython>=6.1.0->ipywidgets) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/mamba/lib/python3.10/site-packages (from pexpect>4.3->ipython>=6.1.0->ipywidgets) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /opt/mamba/lib/python3.10/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython>=6.1.0->ipywidgets) (0.2.10)\n",
      "Requirement already satisfied: executing>=1.2.0 in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets) (2.0.1)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets) (0.2.2)\n",
      "Requirement already satisfied: six>=1.12.0 in /opt/mamba/lib/python3.10/site-packages (from asttokens>=2.1.0->stack-data->ipython>=6.1.0->ipywidgets) (1.16.0)\n",
      "Building wheels for collected packages: nglview\n",
      "  Building wheel for nglview (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for nglview: filename=nglview-3.0.8-py3-none-any.whl size=10216160 sha256=4066092451d863b74695de8137d52600cc1840802a5bd678c932a33a65777166\n",
      "  Stored in directory: /home/onyxia/.cache/pip/wheels/2e/6c/59/32bf4aa0134f9c4cdca054f5192839fb4285241e2f17f7d358\n",
      "Successfully built nglview\n",
      "Installing collected packages: biopython, nglview\n",
      "Successfully installed biopython-1.81 nglview-3.0.8\n"
     ]
    }
   ],
   "source": [
    "!pip install --upgrade biopython nglview ipywidgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cba2fd23-069e-4f9c-96a1-27534ab4b2ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found existing installation: nglview 3.0.8\n",
      "Uninstalling nglview-3.0.8:\n",
      "  Successfully uninstalled nglview-3.0.8\n",
      "Collecting nglview\n",
      "  Using cached nglview-3.0.8-py3-none-any.whl\n",
      "Requirement already satisfied: ipywidgets>=7 in /opt/mamba/lib/python3.10/site-packages (from nglview) (8.1.1)\n",
      "Requirement already satisfied: jupyterlab-widgets in /opt/mamba/lib/python3.10/site-packages (from nglview) (3.0.9)\n",
      "Requirement already satisfied: numpy in /opt/mamba/lib/python3.10/site-packages (from nglview) (1.26.0)\n",
      "Requirement already satisfied: comm>=0.1.3 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets>=7->nglview) (0.1.4)\n",
      "Requirement already satisfied: ipython>=6.1.0 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets>=7->nglview) (8.17.2)\n",
      "Requirement already satisfied: traitlets>=4.3.1 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets>=7->nglview) (5.13.0)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0.9 in /opt/mamba/lib/python3.10/site-packages (from ipywidgets>=7->nglview) (4.0.9)\n",
      "Requirement already satisfied: decorator in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (0.19.1)\n",
      "Requirement already satisfied: matplotlib-inline in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (0.1.6)\n",
      "Requirement already satisfied: prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (3.0.41)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (2.17.0)\n",
      "Requirement already satisfied: stack-data in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (0.6.2)\n",
      "Requirement already satisfied: exceptiongroup in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (1.1.3)\n",
      "Requirement already satisfied: pexpect>4.3 in /opt/mamba/lib/python3.10/site-packages (from ipython>=6.1.0->ipywidgets>=7->nglview) (4.8.0)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /opt/mamba/lib/python3.10/site-packages (from jedi>=0.16->ipython>=6.1.0->ipywidgets>=7->nglview) (0.8.3)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /opt/mamba/lib/python3.10/site-packages (from pexpect>4.3->ipython>=6.1.0->ipywidgets>=7->nglview) (0.7.0)\n",
      "Requirement already satisfied: wcwidth in /opt/mamba/lib/python3.10/site-packages (from prompt-toolkit!=3.0.37,<3.1.0,>=3.0.30->ipython>=6.1.0->ipywidgets>=7->nglview) (0.2.10)\n",
      "Requirement already satisfied: executing>=1.2.0 in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=7->nglview) (2.0.1)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=7->nglview) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /opt/mamba/lib/python3.10/site-packages (from stack-data->ipython>=6.1.0->ipywidgets>=7->nglview) (0.2.2)\n",
      "Requirement already satisfied: six>=1.12.0 in /opt/mamba/lib/python3.10/site-packages (from asttokens>=2.1.0->stack-data->ipython>=6.1.0->ipywidgets>=7->nglview) (1.16.0)\n",
      "Installing collected packages: nglview\n",
      "Successfully installed nglview-3.0.8\n"
     ]
    }
   ],
   "source": [
    "!pip uninstall nglview -y\n",
    "!pip install nglview\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56853dfb-b045-4670-ad5a-f87f8d2c8efe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e8139ba1a22b4cdd96c0f66ab7226c34",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": []
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import ipywidgets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9cbabdec-0975-48a3-8b4f-dbf057cb4bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "cif_parser = MMCIFParser()\n",
    "structure = cif_parser.get_structure(\"6EBK\", \"/home/onyxia/work/TDA_Protein3D/data/AF-P04637-F1-model_v4.cif\")\n",
    "view = nv.show_biopython(structure)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2cf02fa9-cd52-4737-bf18-8fde193f0f44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bbf009772bd446dd8d4d31217ac75a73",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f2c75f8-02e4-47a4-aa0e-e8ef6752c130",
   "metadata": {},
   "source": [
    "Selecting only relevant KO's for TDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "006e5e1a-9950-4460-95ce-9c071492d059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "459"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"/home/onyxia/work/TDA_Protein3D/KO_shap.txt\", \"r\") as ko_to_fetch:\n",
    "    KO_s = ko_to_fetch.read().replace(\" \", \"\").split('\\n')\n",
    "KO_s\n",
    "\n",
    "len(KO_s) #18*30 -> cohérent\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "42512eca-a0a0-4329-a608-287854491522",
   "metadata": {},
   "outputs": [],
   "source": [
    "KO= set(KO_s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3a498d96-eb92-4e65-b702-51c485d52b57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "459"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(KO)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7eb9ed40-74b7-4ecb-9e5d-726d1469e06a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'K00002',\n",
       " 'K00003',\n",
       " 'K00004',\n",
       " 'K00005',\n",
       " 'K00006',\n",
       " 'K00008',\n",
       " 'K00010',\n",
       " 'K00027',\n",
       " 'K00028',\n",
       " 'K00033',\n",
       " 'K00042',\n",
       " 'K00057',\n",
       " 'K00059',\n",
       " 'K00060',\n",
       " 'K00067',\n",
       " 'K00113',\n",
       " 'K00122',\n",
       " 'K00128',\n",
       " 'K00131',\n",
       " 'K00145',\n",
       " 'K00147',\n",
       " 'K00156',\n",
       " 'K00161',\n",
       " 'K00162',\n",
       " 'K00169',\n",
       " 'K00170',\n",
       " 'K00172',\n",
       " 'K00208',\n",
       " 'K00218',\n",
       " 'K00228',\n",
       " 'K00230',\n",
       " 'K00236',\n",
       " 'K00245',\n",
       " 'K00246',\n",
       " 'K00254',\n",
       " 'K00274',\n",
       " 'K00275',\n",
       " 'K00303',\n",
       " 'K00313',\n",
       " 'K00320',\n",
       " 'K00331',\n",
       " 'K00363',\n",
       " 'K00371',\n",
       " 'K00392',\n",
       " 'K00395',\n",
       " 'K00412',\n",
       " 'K00464',\n",
       " 'K00474',\n",
       " 'K00524',\n",
       " 'K00547',\n",
       " 'K00549',\n",
       " 'K00554',\n",
       " 'K00555',\n",
       " 'K00557',\n",
       " 'K00558',\n",
       " 'K00565',\n",
       " 'K00566',\n",
       " 'K00567',\n",
       " 'K00574',\n",
       " 'K00575',\n",
       " 'K00613',\n",
       " 'K00631',\n",
       " 'K00632',\n",
       " 'K00639',\n",
       " 'K00657',\n",
       " 'K00661',\n",
       " 'K00691',\n",
       " 'K00698',\n",
       " 'K00706',\n",
       " 'K00822',\n",
       " 'K00855',\n",
       " 'K00922',\n",
       " 'K00927',\n",
       " 'K00940',\n",
       " 'K00944',\n",
       " 'K00950',\n",
       " 'K00988',\n",
       " 'K00997',\n",
       " 'K00999',\n",
       " 'K01029',\n",
       " 'K01033',\n",
       " 'K01056',\n",
       " 'K01057',\n",
       " 'K01060',\n",
       " 'K01103',\n",
       " 'K01126',\n",
       " 'K01129',\n",
       " 'K01146',\n",
       " 'K01150',\n",
       " 'K01151',\n",
       " 'K01170',\n",
       " 'K01174',\n",
       " 'K01208',\n",
       " 'K01214',\n",
       " 'K01251',\n",
       " 'K01255',\n",
       " 'K01429',\n",
       " 'K01465',\n",
       " 'K01515',\n",
       " 'K01643',\n",
       " 'K01644',\n",
       " 'K01655',\n",
       " 'K01667',\n",
       " 'K01669',\n",
       " 'K01679',\n",
       " 'K01705',\n",
       " 'K01740',\n",
       " 'K01788',\n",
       " 'K01867',\n",
       " 'K01870',\n",
       " 'K01895',\n",
       " 'K01907',\n",
       " 'K01948',\n",
       " 'K01952',\n",
       " 'K01963',\n",
       " 'K02017',\n",
       " 'K02048',\n",
       " 'K02073',\n",
       " 'K02110',\n",
       " 'K02132',\n",
       " 'K02183',\n",
       " 'K02189',\n",
       " 'K02191',\n",
       " 'K02201',\n",
       " 'K02208',\n",
       " 'K02221',\n",
       " 'K02258',\n",
       " 'K02295',\n",
       " 'K02315',\n",
       " 'K02323',\n",
       " 'K02327',\n",
       " 'K02451',\n",
       " 'K02476',\n",
       " 'K02637',\n",
       " 'K02639',\n",
       " 'K02640',\n",
       " 'K02641',\n",
       " 'K02662',\n",
       " 'K02689',\n",
       " 'K02690',\n",
       " 'K02692',\n",
       " 'K02716',\n",
       " 'K02725',\n",
       " 'K02737',\n",
       " 'K02823',\n",
       " 'K02888',\n",
       " 'K02909',\n",
       " 'K02912',\n",
       " 'K02929',\n",
       " 'K02954',\n",
       " 'K02984',\n",
       " 'K03025',\n",
       " 'K03049',\n",
       " 'K03151',\n",
       " 'K03166',\n",
       " 'K03167',\n",
       " 'K03177',\n",
       " 'K03188',\n",
       " 'K03214',\n",
       " 'K03234',\n",
       " 'K03242',\n",
       " 'K03249',\n",
       " 'K03336',\n",
       " 'K03428',\n",
       " 'K03433',\n",
       " 'K03455',\n",
       " 'K03515',\n",
       " 'K03525',\n",
       " 'K03526',\n",
       " 'K03527',\n",
       " 'K03528',\n",
       " 'K03531',\n",
       " 'K03535',\n",
       " 'K03536',\n",
       " 'K03545',\n",
       " 'K03546',\n",
       " 'K03547',\n",
       " 'K03548',\n",
       " 'K03550',\n",
       " 'K03555',\n",
       " 'K03622',\n",
       " 'K03626',\n",
       " 'K03634',\n",
       " 'K03649',\n",
       " 'K03650',\n",
       " 'K03656',\n",
       " 'K03660',\n",
       " 'K03665',\n",
       " 'K03666',\n",
       " 'K03669',\n",
       " 'K03670',\n",
       " 'K03679',\n",
       " 'K03681',\n",
       " 'K03683',\n",
       " 'K03684',\n",
       " 'K03685',\n",
       " 'K03686',\n",
       " 'K03687',\n",
       " 'K03688',\n",
       " 'K03690',\n",
       " 'K03694',\n",
       " 'K03695',\n",
       " 'K03718',\n",
       " 'K03719',\n",
       " 'K03723',\n",
       " 'K03724',\n",
       " 'K03725',\n",
       " 'K03726',\n",
       " 'K03727',\n",
       " 'K03732',\n",
       " 'K03733',\n",
       " 'K03734',\n",
       " 'K03737',\n",
       " 'K03738',\n",
       " 'K03743',\n",
       " 'K03744',\n",
       " 'K03745',\n",
       " 'K03746',\n",
       " 'K03747',\n",
       " 'K03749',\n",
       " 'K03750',\n",
       " 'K03752',\n",
       " 'K03754',\n",
       " 'K03759',\n",
       " 'K03761',\n",
       " 'K03764',\n",
       " 'K03767',\n",
       " 'K03769',\n",
       " 'K03770',\n",
       " 'K03771',\n",
       " 'K03772',\n",
       " 'K03773',\n",
       " 'K03774',\n",
       " 'K03775',\n",
       " 'K03776',\n",
       " 'K03777',\n",
       " 'K03778',\n",
       " 'K03779',\n",
       " 'K03781',\n",
       " 'K03782',\n",
       " 'K03784',\n",
       " 'K03785',\n",
       " 'K03788',\n",
       " 'K03798',\n",
       " 'K03799',\n",
       " 'K03803',\n",
       " 'K03805',\n",
       " 'K03807',\n",
       " 'K03808',\n",
       " 'K03845',\n",
       " 'K03852',\n",
       " 'K03883',\n",
       " 'K03887',\n",
       " 'K03892',\n",
       " 'K03897',\n",
       " 'K03919',\n",
       " 'K03923',\n",
       " 'K03926',\n",
       " 'K03928',\n",
       " 'K03929',\n",
       " 'K03931',\n",
       " 'K03946',\n",
       " 'K03949',\n",
       " 'K03950',\n",
       " 'K03952',\n",
       " 'K03953',\n",
       " 'K03955',\n",
       " 'K03963',\n",
       " 'K03969',\n",
       " 'K03970',\n",
       " 'K03974',\n",
       " 'K04035',\n",
       " 'K04038',\n",
       " 'K04072',\n",
       " 'K04078',\n",
       " 'K04084',\n",
       " 'K04085',\n",
       " 'K04101',\n",
       " 'K04338',\n",
       " 'K04343',\n",
       " 'K04345',\n",
       " 'K04348',\n",
       " 'K04354',\n",
       " 'K04364',\n",
       " 'K04370',\n",
       " 'K04411',\n",
       " 'K04412',\n",
       " 'K04427',\n",
       " 'K04437',\n",
       " 'K04456',\n",
       " 'K04459',\n",
       " 'K04461',\n",
       " 'K04477',\n",
       " 'K04484',\n",
       " 'K04486',\n",
       " 'K04487',\n",
       " 'K04488',\n",
       " 'K04494',\n",
       " 'K04496',\n",
       " 'K04498',\n",
       " 'K04499',\n",
       " 'K04505',\n",
       " 'K04507',\n",
       " 'K04508',\n",
       " 'K04513',\n",
       " 'K04515',\n",
       " 'K04516',\n",
       " 'K04517',\n",
       " 'K04518',\n",
       " 'K04522',\n",
       " 'K04523',\n",
       " 'K04532',\n",
       " 'K04534',\n",
       " 'K04536',\n",
       " 'K04554',\n",
       " 'K04555',\n",
       " 'K04560',\n",
       " 'K04561',\n",
       " 'K04562',\n",
       " 'K04564',\n",
       " 'K04565',\n",
       " 'K04566',\n",
       " 'K04567',\n",
       " 'K04568',\n",
       " 'K04569',\n",
       " 'K04615',\n",
       " 'K04618',\n",
       " 'K04630',\n",
       " 'K04631',\n",
       " 'K04651',\n",
       " 'K04652',\n",
       " 'K04653',\n",
       " 'K04654',\n",
       " 'K04655',\n",
       " 'K04656',\n",
       " 'K04678',\n",
       " 'K04681',\n",
       " 'K04683',\n",
       " 'K04688',\n",
       " 'K04702',\n",
       " 'K04719',\n",
       " 'K04720',\n",
       " 'K04768',\n",
       " 'K04957',\n",
       " 'K05377',\n",
       " 'K05524',\n",
       " 'K05570',\n",
       " 'K05576',\n",
       " 'K05770',\n",
       " 'K05827',\n",
       " 'K05831',\n",
       " 'K05921',\n",
       " 'K05929',\n",
       " 'K05934',\n",
       " 'K05942',\n",
       " 'K06446',\n",
       " 'K06689',\n",
       " 'K06871',\n",
       " 'K06876',\n",
       " 'K06955',\n",
       " 'K06982',\n",
       " 'K06988',\n",
       " 'K06990',\n",
       " 'K07005',\n",
       " 'K07023',\n",
       " 'K07040',\n",
       " 'K07060',\n",
       " 'K07080',\n",
       " 'K07448',\n",
       " 'K07507',\n",
       " 'K07558',\n",
       " 'K07572',\n",
       " 'K07583',\n",
       " 'K07588',\n",
       " 'K07642',\n",
       " 'K07739',\n",
       " 'K07750',\n",
       " 'K07755',\n",
       " 'K07760',\n",
       " 'K08193',\n",
       " 'K08310',\n",
       " 'K08676',\n",
       " 'K08685',\n",
       " 'K08691',\n",
       " 'K08745',\n",
       " 'K08776',\n",
       " 'K08903',\n",
       " 'K09023',\n",
       " 'K09251',\n",
       " 'K09728',\n",
       " 'K09738',\n",
       " 'K09764',\n",
       " 'K09771',\n",
       " 'K09779',\n",
       " 'K09798',\n",
       " 'K09812',\n",
       " 'K09866',\n",
       " 'K09919',\n",
       " 'K10018',\n",
       " 'K10118',\n",
       " 'K10223',\n",
       " 'K10232',\n",
       " 'K10234',\n",
       " 'K10671',\n",
       " 'K10807',\n",
       " 'K10823',\n",
       " 'K10914',\n",
       " 'K11529',\n",
       " 'K11727',\n",
       " 'K11751',\n",
       " 'K11754',\n",
       " 'K11921',\n",
       " 'K12064',\n",
       " 'K12666',\n",
       " 'K13013',\n",
       " 'K13503',\n",
       " 'K13574',\n",
       " 'K13626',\n",
       " 'K13995',\n",
       " 'K14165',\n",
       " 'K14534',\n",
       " 'K15229',\n",
       " 'K15429',\n",
       " 'K15449',\n",
       " 'K15635',\n",
       " 'K15865',\n",
       " 'K16345',\n",
       " 'K16710',\n",
       " 'K16740',\n",
       " 'K16885',\n",
       " 'K16887',\n",
       " 'K17050',\n",
       " 'K17681',\n",
       " 'K17865',\n",
       " 'K18587',\n",
       " 'K18588',\n",
       " 'K18672',\n",
       " 'K18828',\n",
       " 'K18929',\n",
       " 'K19199',\n",
       " 'K19511',\n",
       " 'K19736',\n",
       " 'K19745',\n",
       " 'K19882',\n",
       " 'K20305',\n",
       " 'K21480',\n",
       " 'K21703',\n",
       " 'K21801',\n",
       " 'K21947',\n",
       " 'K22024',\n",
       " 'K22616',\n",
       " 'K22699',\n",
       " 'K22757',\n",
       " 'K22761',\n",
       " 'K22770',\n",
       " 'k00281',\n",
       " 'К03746',\n",
       " 'К04460'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41b93704-bb2c-47e6-8ffd-0f57f64ee15a",
   "metadata": {},
   "source": [
    "Downloading the data from Luis google drive (preferably with ethernet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ca742734-7e6f-4d86-bafd-a9eddca21f82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-11-28 14:59:45--  https://docs.google.com/uc?export=download&confirm=t&id=1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s\n",
      "Resolving docs.google.com (docs.google.com)... 142.250.179.78, 2a00:1450:4007:810::200e\n",
      "Connecting to docs.google.com (docs.google.com)|142.250.179.78|:443... connected.\n",
      "HTTP request sent, awaiting response... 303 See Other\n",
      "Location: https://doc-0k-90-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/laqatm9f6eiaiqufedmqjik5k4ua1r61/1701183525000/01727642121889718427/*/1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s?e=download&uuid=86aa8046-b6ea-43d1-8dfe-8f683fd5360e [following]\n",
      "Warning: wildcards not supported in HTTP.\n",
      "--2023-11-28 14:59:45--  https://doc-0k-90-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/laqatm9f6eiaiqufedmqjik5k4ua1r61/1701183525000/01727642121889718427/*/1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s?e=download&uuid=86aa8046-b6ea-43d1-8dfe-8f683fd5360e\n",
      "Resolving doc-0k-90-docs.googleusercontent.com (doc-0k-90-docs.googleusercontent.com)... 142.250.179.65, 2a00:1450:4007:80e::2001\n",
      "Connecting to doc-0k-90-docs.googleusercontent.com (doc-0k-90-docs.googleusercontent.com)|142.250.179.65|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 35575974570 (33G) [text/tab-separated-values]\n",
      "Saving to: ‘OM-RGC_v2.tsv’\n",
      "\n",
      "OM-RGC_v2.tsv       100%[===================>]  33.13G  21.9MB/s    in 24m 51s \n",
      "\n",
      "2023-11-28 15:24:36 (22.8 MB/s) - ‘OM-RGC_v2.tsv’ saved [35575974570/35575974570]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# OM-RGC_v2.tsv https://drive.google.com/file/d/1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s/view?usp=sharing\n",
    "# To download gene_profile_metaG.tsv from my inria account\n",
    "!wget --load-cookies /tmp/cookies.txt \"https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\\1\\n/p')&id=1YUNHJXFJ3T8bcoi1YEI8gZ5e6mBzY93s\" -O OM-RGC_v2.tsv && rm -rf /tmp/cookies.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "16d9845c-a6e4-4200-98a2-693b2053d01b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "\n",
    "# Création d'une requête\n",
    "q = (\n",
    "    pl.scan_csv(\"/home/onyxia/work/TDA_Protein3D/notebooks/OM-RGC_v2.tsv\", separator='\\t') # Lecture lazy\n",
    "    .filter(pl.col(\"KO\").is_in(KO))\n",
    "    .with_columns(pl.col('sequence').str.starts_with('ATG').alias('Prot'))\n",
    ")\n",
    "\n",
    "# Exécution de la requête\n",
    "df = q.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ab859eba-2be2-4fb1-bb35-d09f95110d98",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (995_318, 15)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>gene</th><th>OM-RGC_ID</th><th>KO</th><th>OG</th><th>GC</th><th>Domain</th><th>Phylum</th><th>Class</th><th>Order</th><th>Family</th><th>Genus</th><th>Species</th><th>Strain</th><th>sequence</th><th>Prot</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>bool</td></tr></thead><tbody><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K03531&quot;</td><td>&quot;COG0206&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>null</td><td>null</td><td>null</td><td>&quot;AACTTTAAAACACC…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K00147&quot;</td><td>&quot;COG0014&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>&quot;Candidatus Pel…</td><td>null</td><td>null</td><td>&quot;CCACTTGGCGTTAT…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K03545&quot;</td><td>&quot;COG0544&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Gammaproteobac…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;AATATCATTCAGAG…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0267…</td><td>&quot;K01895&quot;</td><td>&quot;COG0365&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;GAAACTTATTTTTC…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K01907&quot;</td><td>&quot;COG0365&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>&quot;Candidatus Pel…</td><td>null</td><td>null</td><td>&quot;AATCAAAAAGGGGA…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0260…</td><td>&quot;K00254&quot;</td><td>&quot;COG0167&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>&quot;Candidatus Pel…</td><td>&quot;Candidatus Pel…</td><td>null</td><td>&quot;CTTAAACTTAATTT…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K00010&quot;</td><td>&quot;COG0673&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>null</td><td>null</td><td>null</td><td>&quot;CCCGCTCCTCCTTC…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0298…</td><td>&quot;K00228&quot;</td><td>&quot;COG0408&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>&quot;Candidatus Pel…</td><td>null</td><td>null</td><td>&quot;GTATTAAGGATGAT…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K03336&quot;</td><td>&quot;COG3962&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>null</td><td>null</td><td>null</td><td>&quot;TCTTTTAAAGCTGT…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0277…</td><td>&quot;K00254&quot;</td><td>&quot;COG0167&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>null</td><td>&quot;SAR11 cluster …</td><td>null</td><td>&quot;TTGAATAAACTAAA…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0263…</td><td>&quot;K01465&quot;</td><td>&quot;COG0044&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;AATCAATCAATTGG…</td><td>false</td></tr><tr><td>&quot;TARA_X00000036…</td><td>&quot;OM-RGC.v2.0335…</td><td>&quot;K10234&quot;</td><td>&quot;COG0395&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Alphaproteobac…</td><td>&quot;Pelagibacteral…</td><td>&quot;Pelagibacterac…</td><td>null</td><td>&quot;alpha proteoba…</td><td>null</td><td>&quot;AAATCACTACCCCA…</td><td>false</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0176…</td><td>&quot;K06446&quot;</td><td>&quot;COG1960&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;ATAAAAGGCGCACG…</td><td>false</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0250…</td><td>&quot;K06446&quot;</td><td>&quot;COG1960&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;ATGATTCGTGACCA…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0077…</td><td>&quot;K03773&quot;</td><td>&quot;COG0545&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Gammaproteobac…</td><td>&quot;Pseudomonadale…</td><td>&quot;Moraxellaceae&quot;</td><td>&quot;Perlucidibaca&quot;</td><td>null</td><td>null</td><td>&quot;ATGGTTTATGCAAA…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0160…</td><td>&quot;K03634&quot;</td><td>&quot;COG2834&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Gammaproteobac…</td><td>&quot;Oceanospirilla…</td><td>&quot;Oleiphilaceae&quot;</td><td>&quot;Oleiphilus&quot;</td><td>null</td><td>null</td><td>&quot;ATGAATAAAACAAT…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0043…</td><td>&quot;K03688&quot;</td><td>&quot;COG0661&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Gammaproteobac…</td><td>&quot;Alteromonadale…</td><td>&quot;Shewanellaceae…</td><td>&quot;Shewanella&quot;</td><td>&quot;Shewanella fri…</td><td>&quot;Shewanella fri…</td><td>&quot;ATGCTTTGGGAAAC…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0019…</td><td>&quot;K03726&quot;</td><td>&quot;COG1204&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Candidatus Hyd…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;Candidatus Hyd…</td><td>null</td><td>&quot;GAAATCCTCAGCGA…</td><td>false</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0181…</td><td>&quot;K00156&quot;</td><td>&quot;COG0028&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Bacteroidetes&quot;</td><td>&quot;Flavobacteriia…</td><td>&quot;Flavobacterial…</td><td>&quot;Flavobacteriac…</td><td>null</td><td>null</td><td>null</td><td>&quot;GATGCCATATTTAC…</td><td>false</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0065…</td><td>&quot;K00156&quot;</td><td>&quot;COG0028&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>&quot;Gammaproteobac…</td><td>&quot;Alteromonadale…</td><td>&quot;Shewanellaceae…</td><td>&quot;Shewanella&quot;</td><td>null</td><td>null</td><td>&quot;ATGTCAAAAAATAT…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0218…</td><td>&quot;K00940&quot;</td><td>&quot;COG0105&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>&quot;Proteobacteria…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;AACGGCATTGGCAG…</td><td>false</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0169…</td><td>&quot;K01895&quot;</td><td>&quot;COG0365&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;GAAGCCTGGATCTG…</td><td>false</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0046…</td><td>&quot;K01895&quot;</td><td>&quot;COG0365&quot;</td><td>null</td><td>&quot;Bacteria&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;ATGAGTCAACAGAT…</td><td>true</td></tr><tr><td>&quot;TARA_R11000159…</td><td>&quot;OM-RGC.v2.0130…</td><td>&quot;K01895&quot;</td><td>&quot;COG0365&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;TATAACAACTTTGA…</td><td>false</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (995_318, 15)\n",
       "┌────────────────┬───────────────┬────────┬─────────┬───┬─────────┬────────┬───────────────┬───────┐\n",
       "│ gene           ┆ OM-RGC_ID     ┆ KO     ┆ OG      ┆ … ┆ Species ┆ Strain ┆ sequence      ┆ Prot  │\n",
       "│ ---            ┆ ---           ┆ ---    ┆ ---     ┆   ┆ ---     ┆ ---    ┆ ---           ┆ ---   │\n",
       "│ str            ┆ str           ┆ str    ┆ str     ┆   ┆ str     ┆ str    ┆ str           ┆ bool  │\n",
       "╞════════════════╪═══════════════╪════════╪═════════╪═══╪═════════╪════════╪═══════════════╪═══════╡\n",
       "│ TARA_X00000036 ┆ OM-RGC.v2.026 ┆ K03531 ┆ COG0206 ┆ … ┆ null    ┆ null   ┆ AACTTTAAAACAC ┆ false │\n",
       "│ 8_G_C22366181_ ┆ 366502        ┆        ┆         ┆   ┆         ┆        ┆ CAGAGATAAAAGA ┆       │\n",
       "│ 1_ge…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ GTTACA…       ┆       │\n",
       "│ TARA_X00000036 ┆ OM-RGC.v2.026 ┆ K00147 ┆ COG0014 ┆ … ┆ null    ┆ null   ┆ CCACTTGGCGTTA ┆ false │\n",
       "│ 8_G_C22367915_ ┆ 366588        ┆        ┆         ┆   ┆         ┆        ┆ TTGGTATAATATA ┆       │\n",
       "│ 1_ge…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ TGAAAG…       ┆       │\n",
       "│ TARA_X00000036 ┆ OM-RGC.v2.026 ┆ K03545 ┆ COG0544 ┆ … ┆ null    ┆ null   ┆ AATATCATTCAGA ┆ false │\n",
       "│ 8_G_C22368449_ ┆ 366615        ┆        ┆         ┆   ┆         ┆        ┆ GCCAGTATGGCAA ┆       │\n",
       "│ 1_ge…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ TCAGGT…       ┆       │\n",
       "│ TARA_X00000036 ┆ OM-RGC.v2.026 ┆ K01895 ┆ COG0365 ┆ … ┆ null    ┆ null   ┆ GAAACTTATTTTT ┆ false │\n",
       "│ 8_G_C22369051_ ┆ 732933        ┆        ┆         ┆   ┆         ┆        ┆ CTCAGTTTGATGG ┆       │\n",
       "│ 1_ge…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ AAAGTA…       ┆       │\n",
       "│ …              ┆ …             ┆ …      ┆ …       ┆ … ┆ …       ┆ …      ┆ …             ┆ …     │\n",
       "│ TARA_R11000159 ┆ OM-RGC.v2.021 ┆ K00940 ┆ COG0105 ┆ … ┆ null    ┆ null   ┆ AACGGCATTGGCA ┆ false │\n",
       "│ 2_G_scaffold99 ┆ 842062        ┆        ┆         ┆   ┆         ┆        ┆ GCCGCCAATCGGC ┆       │\n",
       "│ 540_…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ GCGCCA…       ┆       │\n",
       "│ TARA_R11000159 ┆ OM-RGC.v2.016 ┆ K01895 ┆ COG0365 ┆ … ┆ null    ┆ null   ┆ GAAGCCTGGATCT ┆ false │\n",
       "│ 2_G_scaffold99 ┆ 926211        ┆        ┆         ┆   ┆         ┆        ┆ GGTACCACAAGCA ┆       │\n",
       "│ 713_…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ TATTGG…       ┆       │\n",
       "│ TARA_R11000159 ┆ OM-RGC.v2.004 ┆ K01895 ┆ COG0365 ┆ … ┆ null    ┆ null   ┆ ATGAGTCAACAGA ┆ true  │\n",
       "│ 2_G_scaffold99 ┆ 640879        ┆        ┆         ┆   ┆         ┆        ┆ TTGAACACGTCCT ┆       │\n",
       "│ 713_…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ TGAAGA…       ┆       │\n",
       "│ TARA_R11000159 ┆ OM-RGC.v2.013 ┆ K01895 ┆ COG0365 ┆ … ┆ null    ┆ null   ┆ TATAACAACTTTG ┆ false │\n",
       "│ 2_G_scaffold99 ┆ 062172        ┆        ┆         ┆   ┆         ┆        ┆ ATCCCGCCGCTCT ┆       │\n",
       "│ 769_…          ┆               ┆        ┆         ┆   ┆         ┆        ┆ GGAAGC…       ┆       │\n",
       "└────────────────┴───────────────┴────────┴─────────┴───┴─────────┴────────┴───────────────┴───────┘"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c7ff3298-422c-47d1-bd57-2dd960fb36ad",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df.write_parquet(\"/home/onyxia/work/TDA_Protein3D/data/TARA_important_genes.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3039ce27-49f0-4294-b201-4ca44bda9606",
   "metadata": {},
   "source": [
    "Now we want to fetch the structures associated with each genes: we thus either need to query via the RNA sequence or via an external identifier we manage to retrieve via KEGG, UniProt, or any other famous database. Retrieving by such a way would allow us to verify that our gene is indeed a protein (and thus it makes sense to retrieve its 3D structure)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "87ed498e-1458-4cd7-9bee-64f174116d1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "row_2 = df.filter(pl.col('Prot')== True)[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "0eb0a303-c8c9-4f8b-9922-1f2aa4fa00df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGTTAAAAAAATTTTTAGATAACTTTAATTTTTCGTATAAATTTAAAAGTTCAACTGAACTTTACAAATCTGGTTTTTTTAATTCCACACTACAAATTATTTTAGAAAATTATGATGGAATAATGAATATTATATTACCAACATTAGGAAAAGAACGTCAAAAAACTTACAGTCCATTTCTACCACTATGTCCTGAAACTGGTCATGTTTTAGAAATACCAGTTAAGGATATAAATAAAGATAAATCAAAAATTATTTTTGACAATTATGGAAAAGAATTAGAAGTAAGCATTTTAAATGGAAATTGTAAATTACAATGGAAAGTTGATTGGGCCATGAGGTGGTACGCATTAGATATAGATTTTGAAATGTATGGAAAAGATCTTATTGAAAGTGCGATCTTATCAACAAAAATAATTAAGTTAATTGGAAAAAAAAATCCATCTGGATTTGCATATGAATTATTTTTA'"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_2.select(pl.col('sequence')).item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "f08f3584-fc99-4347-8012-6f31535cd1c5",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__add__',\n",
       " '__annotations__',\n",
       " '__array__',\n",
       " '__bool__',\n",
       " '__class__',\n",
       " '__contains__',\n",
       " '__copy__',\n",
       " '__dataframe__',\n",
       " '__dataframe_consortium_standard__',\n",
       " '__deepcopy__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__floordiv__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__getitem__',\n",
       " '__getstate__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__iter__',\n",
       " '__le__',\n",
       " '__len__',\n",
       " '__lt__',\n",
       " '__mod__',\n",
       " '__module__',\n",
       " '__mul__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__radd__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__reversed__',\n",
       " '__rmul__',\n",
       " '__setattr__',\n",
       " '__setitem__',\n",
       " '__setstate__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__sub__',\n",
       " '__subclasshook__',\n",
       " '__truediv__',\n",
       " '__weakref__',\n",
       " '_accessors',\n",
       " '_cast_all_from_to',\n",
       " '_comp',\n",
       " '_compare_to_non_df',\n",
       " '_compare_to_other_df',\n",
       " '_df',\n",
       " '_div',\n",
       " '_from_arrow',\n",
       " '_from_dict',\n",
       " '_from_dicts',\n",
       " '_from_numpy',\n",
       " '_from_pandas',\n",
       " '_from_pydf',\n",
       " '_from_records',\n",
       " '_ipython_key_completions_',\n",
       " '_pos_idx',\n",
       " '_read_avro',\n",
       " '_read_csv',\n",
       " '_read_ipc',\n",
       " '_read_ipc_stream',\n",
       " '_read_json',\n",
       " '_read_ndjson',\n",
       " '_read_parquet',\n",
       " '_replace',\n",
       " '_repr_html_',\n",
       " '_take_with_series',\n",
       " 'apply',\n",
       " 'approx_n_unique',\n",
       " 'approx_unique',\n",
       " 'bottom_k',\n",
       " 'cast',\n",
       " 'clear',\n",
       " 'clone',\n",
       " 'columns',\n",
       " 'corr',\n",
       " 'describe',\n",
       " 'drop',\n",
       " 'drop_in_place',\n",
       " 'drop_nulls',\n",
       " 'dtypes',\n",
       " 'estimated_size',\n",
       " 'explode',\n",
       " 'extend',\n",
       " 'fill_nan',\n",
       " 'fill_null',\n",
       " 'filter',\n",
       " 'find_idx_by_name',\n",
       " 'flags',\n",
       " 'fold',\n",
       " 'frame_equal',\n",
       " 'gather_every',\n",
       " 'get_column',\n",
       " 'get_column_index',\n",
       " 'get_columns',\n",
       " 'glimpse',\n",
       " 'group_by',\n",
       " 'group_by_dynamic',\n",
       " 'group_by_rolling',\n",
       " 'groupby',\n",
       " 'groupby_dynamic',\n",
       " 'groupby_rolling',\n",
       " 'hash_rows',\n",
       " 'head',\n",
       " 'height',\n",
       " 'hstack',\n",
       " 'insert_at_idx',\n",
       " 'insert_column',\n",
       " 'interpolate',\n",
       " 'is_duplicated',\n",
       " 'is_empty',\n",
       " 'is_unique',\n",
       " 'item',\n",
       " 'iter_rows',\n",
       " 'iter_slices',\n",
       " 'join',\n",
       " 'join_asof',\n",
       " 'lazy',\n",
       " 'limit',\n",
       " 'map_rows',\n",
       " 'max',\n",
       " 'max_horizontal',\n",
       " 'mean',\n",
       " 'mean_horizontal',\n",
       " 'median',\n",
       " 'melt',\n",
       " 'merge_sorted',\n",
       " 'min',\n",
       " 'min_horizontal',\n",
       " 'n_chunks',\n",
       " 'n_unique',\n",
       " 'null_count',\n",
       " 'partition_by',\n",
       " 'pipe',\n",
       " 'pivot',\n",
       " 'product',\n",
       " 'quantile',\n",
       " 'rechunk',\n",
       " 'rename',\n",
       " 'replace',\n",
       " 'replace_at_idx',\n",
       " 'replace_column',\n",
       " 'reverse',\n",
       " 'rolling',\n",
       " 'row',\n",
       " 'rows',\n",
       " 'rows_by_key',\n",
       " 'sample',\n",
       " 'schema',\n",
       " 'select',\n",
       " 'select_seq',\n",
       " 'set_sorted',\n",
       " 'shape',\n",
       " 'shift',\n",
       " 'shift_and_fill',\n",
       " 'shrink_to_fit',\n",
       " 'slice',\n",
       " 'sort',\n",
       " 'std',\n",
       " 'sum',\n",
       " 'sum_horizontal',\n",
       " 'tail',\n",
       " 'take_every',\n",
       " 'to_arrow',\n",
       " 'to_dict',\n",
       " 'to_dicts',\n",
       " 'to_dummies',\n",
       " 'to_init_repr',\n",
       " 'to_numpy',\n",
       " 'to_pandas',\n",
       " 'to_series',\n",
       " 'to_struct',\n",
       " 'top_k',\n",
       " 'transpose',\n",
       " 'unique',\n",
       " 'unnest',\n",
       " 'unstack',\n",
       " 'update',\n",
       " 'upsample',\n",
       " 'var',\n",
       " 'vstack',\n",
       " 'width',\n",
       " 'with_columns',\n",
       " 'with_columns_seq',\n",
       " 'with_row_count',\n",
       " 'write_avro',\n",
       " 'write_csv',\n",
       " 'write_database',\n",
       " 'write_delta',\n",
       " 'write_excel',\n",
       " 'write_ipc',\n",
       " 'write_ipc_stream',\n",
       " 'write_json',\n",
       " 'write_ndjson',\n",
       " 'write_parquet']"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(row_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "13b63fcb-689d-4e9a-89b7-cfdca52f76ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# Step 1: Translate DNA sequences to protein sequences\n",
    "def translate_dna_to_protein(dna_sequence):\n",
    "    dna_seq = Seq(dna_sequence)\n",
    "    protein_seq = dna_seq.translate()\n",
    "    return str(protein_seq)\n",
    "\n",
    "# Step 2: Search UniProt with protein sequences\n",
    "def search_uniprot(protein_sequence):\n",
    "    result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", protein_sequence)\n",
    "    return NCBIXML.read(result_handle)\n",
    "\n",
    "# Step 3: Extract UniProt accession IDs from the search results\n",
    "def extract_uniprot_accessions(blast_results):\n",
    "    accession_ids = []\n",
    "    for alignment in blast_results.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            accession_ids.append(alignment.accession)\n",
    "    return accession_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a3796801-6ebe-4532-81b6-fd7c6a34ed7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_sequence = 'ATGTTAAAAAAATTTTTAGATAACTTTAATTTTTCGTATAAATTTAAAAGTTCAACTGAACTTTACAAATCTGGTTTTTTTAATTCCACACTACAAATTATTTTAGAAAATTATGATGGAATAATGAATATTATATTACCAACATTAGGAAAAGAACGTCAAAAAACTTACAGTCCATTTCTACCACTATGTCCTGAAACTGGTCATGTTTTAGAAATACCAGTTAAGGATATAAATAAAGATAAATCAAAAATTATTTTTGACAATTATGGAAAAGAATTAGAAGTAAGCATTTTAAATGGAAATTGTAAATTACAATGGAAAGTTGATTGGGCCATGAGGTGGTACGCATTAGATATAGATTTTGAAATGTATGGAAAAGATCTTATTGAAAGTGCGATCTTATCAACAAAAATAATTAAGTTAATTGGAAAAAAAAATCCATCTGGATTTGCATATGAATTATTTTTA' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f3775edd-0db9-40b4-995b-3defc172f8d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MLKKFLDNFNFSYKFKSSTELYKSGFFNSTLQIILENYDGIMNIILPTLGKERQKTYSPFLPLCPETGHVLEIPVKDINKDKSKIIFDNYGKELEVSILNGNCKLQWKVDWAMRWYALDIDFEMYGKDLIESAILSTKIIKLIGKKNPSGFAYELFL\n",
      "UniProt Accession IDs:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "protein_sequence = str(Seq(dna_sequence).translate())\n",
    "print(protein_sequence)\n",
    "# Step 2: Search UniProt with protein sequences\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"uniprotkb\", protein_sequence)\n",
    "blast_results = NCBIXML.read(result_handle)\n",
    "\n",
    "# Step 3: Extract UniProt accession IDs from the search results\n",
    "accession_ids = []\n",
    "for alignment in blast_results.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        accession_ids.append(alignment.accession)\n",
    "\n",
    "# Display the UniProt accession IDs\n",
    "print(\"UniProt Accession IDs:\")\n",
    "print(accession_ids)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4aeda905-6e4f-4284-9220-9efb79e74d7f",
   "metadata": {},
   "source": [
    "### Requests to ESM datafold "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2f8f5048-f32a-49cb-b5b1-cdd8f2948664",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from requests.packages.urllib3.exceptions import InsecureRequestWarning\n",
    "\n",
    "# Disable SSL warnings\n",
    "requests.packages.urllib3.disable_warnings(InsecureRequestWarning)\n",
    "seq = 'GENGEIPLEIRATTGAEVDTRAVTAVEMTEGTLGIFRLPEEDYTALENFRYNRVAGENWKPASTVIYVGGTYARLCAYAPYNSVEFKNSSLKTEAGLTMQTYAAEKDMRFAVSGGDEVWKKTPTANFELKRAYARLVLSVVRDATYPNTCKITKAKIEAFTGNIITANTVDISTGTEGSGTQTPQYIHTVTTGLKDGFAIGLPQQTFSGGVVLTLTVDGMEYSVTIPANKLSTFVRGTKYIVSLAVKGGKLTLMSDKILIDKDWAEVQTGTGGSGDDYDTSFN'\n",
    "url = \"https://api.esmatlas.com/foldSequence/v1/pdb/\"\n",
    "data = {\n",
    "    \"SEQUENCE\" : seq\n",
    "}\n",
    "\n",
    "# Disable SSL certificate verification\n",
    "response = requests.post(url, data=data, verify=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6fa67d39-9c3c-4947-83a8-42c3f37edf44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence is only allowed to have {'H', 'S', 'G', 'W', 'X', 'L', 'D', 'T', 'A', 'J', 'P', 'Z', 'B', 'I', 'K', 'Y', 'M', 'V', 'N', 'R', 'C', 'E', 'Q', 'F'} tokens.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "url = \"https://api.esmatlas.com/foldSequence/v1/pdb/\"\n",
    "\n",
    "data = {\n",
    "    \"SEQUENCE\": \"GENGEIPLEIRATTGAEVDTRAVTAVEMTEGTLGIFRLPEEDYTALENFRYNRVAGENWKPASTVIYVGGTYARLCAYAPYNSVEFKNSSLKTEAGLTMQTYAAEKDMRFAVSGGDEVWKKTPTANFELKRAYARLVLSVVRDATYPNTCKITKAKIEAFTGNIITANTVDISTGTEGSGTQTPQYIHTVTTGLKDGFAIGLPQQTFSGGVVLTLTVDGMEYSVTIPANKLSTFVRGTKYIVSLAVKGGKLTLMSDKILIDKDWAEVQTGTGGSGDDYDTSFN\"\n",
    "}\n",
    "\n",
    "headers = {\n",
    "    \"Content-Type\": \"application/x-www-form-urlencoded\",\n",
    "}\n",
    "\n",
    "response = requests.post(url, data=data, headers=headers, verify=False)\n",
    "\n",
    "print(response.text)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "9b1fb02e-1fa4-419e-b9f1-a769244aff97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Sequence is only allowed to have {'K', 'Y', 'D', 'S', 'W', 'I', 'X', 'T', 'Q', 'V', 'H', 'E', 'N', 'J', 'B', 'C', 'F', 'G', 'R', 'M', 'A', 'P', 'L', 'Z'} tokens.\""
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "67155532-6bcf-4088-bda9-7f784fd2941c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequence = \"MLKKFLDNFNFSYKFKSSTELYKSGFFNSTLQIILENYDGIMNIILPTLGKERQKTYSPFLPLCPETGHVLEIPVKDINKDKSKIIFDNYGKELEVSILNGNCKLQWKVDWAMRWYALDIDFEMYGKDLIESAILSTKIIKLIGKKNPSGFAYELFL\"\n",
    "valid_vocab = ['V', 'C', 'Z', 'R', 'A', 'N', 'X', 'F', 'J', 'Q', 'S', 'B', 'M', 'H', 'I', 'E', 'T', 'Y', 'G', 'W', 'K', 'L', 'D', 'P']\n",
    "for i in list(sequence):\n",
    "    if i not in valid_vocab:\n",
    "        print(i)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
